wallet/storage/src/internal/mod.rs

```diff
@@ -31,13 +37,58 @@ use wallet_types::{
     AccountTxId, KeychainUsageState, RootKeyContent, RootKeyId, WalletTx,
 };
 
+fn password_to_sym_key(password: String, salt: &[u8; 32]) -> SymmetricKey {
+    let hash = argon2id_hash(
+        // TODO: hardcoded values
+        &Argon2Config::new(700, 16, 2),
+        salt,
+        32.try_into().unwrap(),
+        password.as_bytes(),
+    )
+    // TODO: will panic if password.len() > 2^32
+    .expect("correct inputs for hashing");
+
+    SymmetricKey::new_chacha20_from_bytes(hash.try_into().expect("must be able to convert"))
+}
+
 /// Store for wallet data, parametrized over the backend B
-pub struct Store<B: storage::Backend>(storage::Storage<B, Schema>);
+pub struct Store<B: storage::Backend>(storage::Storage<B, Schema>, Option<SymmetricKey>);
 
 impl<B: storage::Backend> Store<B> {
     /// Create a new wallet storage
-    pub fn new(backend: B) -> crate::Result<Self> {
-        let storage = Self(storage::Storage::new(backend).map_err(crate::Error::from)?);
+    pub fn new(backend: B, password: Option<String>) -> crate::Result<Self> {
```

Comment:
I'm becoming more inclined to not put the password on startup at the constructor. This doesn't seem to be correct.

The way I see it is: The wallet open without a password, whether the private keys are encrypted or not. Since we're using xpub, we can derive all the public keys we need without a password. 

To encrypt/decrypt and similar functions, there should be functions in the wallet controller:

- `decrypt(password)`: Opens the wallet, and stores the key in memory, so that the private key can be used
- `encrypt(Option<password>)`: This function encrypts the current wallet, assuming it's decrypted in memory. `None` would remove the encryption.

The functions that require the private key will first check if the wallet is decrypted.

Reply:
Agreed, bitcoin core wallet also does that with lock/unlock. You can also use the wallet as watch only without any private keys.

The difference being they still store all private keys encrypted, and only keep the encryption keys (not sure why they can have multiple) to decrypt the specific private key before usage.
Unlock iterates the map of the private keys and only tries to see if it can decrypt them all and returns an error if not.

Reply:
Yep. Let's call it lock/unlock. That sounds clearer.

Reply:
added 3 separate methods for encrypt/unlock/lock private_keys